### PR TITLE
Add functionality to clear shader

### PR DESCRIPTION
**Pull request by yiqunw700**
1. Add a clear all shader menu to clear all shader data from the figma design file
2. Add a clear shader button to clear the shader from the current selection in the plugin UI

Screenshot 2024-12-18 at 11 39 50

Fixes: #1714

### CODE REVIEW

File: support-figma/extended-layout-plugin/src/shader.ts [hunk 1]
```diff
@@ -131,3 +141,29 @@ export async function setShader(shader: string, shaderFallbackColor: string) {
...
+  Utils.dcLog("Clear shader data from plugin data.");
+
+  await figma.loadAllPagesAsync();
+
+  for (let page of figma.root.children) {
+    for (let child of page.children) {
+      await clearNodeAsync(child);
+    }
+  }
+  figma.closePlugin("Shader data has been cleared.");
```
Comment (rylin8) [inline support-figma/extended-layout-plugin/src/shader.ts:156]: This clears all shaders from the entire file, so maybe something like "All shaders have been removed".

**Review by rylin8 — COMMENTED** [review]
I was thinking that there would be a button next to "Apply Shader" to clear a single node of its shader, instead of, or in addition to, this method of clearing all shaders. A user might have a few shaders in a document and then accidentally apply a shader on the wrong node, in which case there's no way to just remove the shader from that node without clearing all shaders.

Secondly, I don't think there is a way right now to tell if a node has a shader set on it. I think that the shader UI should have some indication that the currently selected node has a shader saved in it, and a way to load that shader code into the text edit box.

File: support-figma/extended-layout-plugin/src/shader.ts (hunk 1 above)
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/shader.ts]: Done.

**Comment by yiqunw700** [comment]
> I was thinking that there would be a button next to "Apply Shader" to clear a single node of its shader, instead of, or in addition to, this method of clearing all shaders. A user might have a few shaders in a document and then accidentally apply a shader on the wrong node, in which case there's no way to just remove the shader from that node without clearing all shaders.
> 
> Secondly, I don't think there is a way right now to tell if a node has a shader set on it. I think that the shader UI should have some indication that the currently selected node has a shader saved in it, and a way to load that shader code into the text edit box.

Good point.
Added a clear shader button in the plugin ui which doesn't check if there is any shader code set though. There is a simple but not very reliable way is to check if the node uses a shader preview image.
filed a ticket #1877  to track the load shader code in the plugin ui.

**Review by rylin8 — APPROVED** [review]

File: support-figma/extended-layout-plugin/src/shader.ts [hunk 2]
```diff
@@ -131,3 +141,66 @@ export async function setShader(shader: string, shaderFallbackColor: string) {
...
+function clearShaderFromNode(node: SceneNode) {
+  let shaderCode = node.getSharedPluginData(
+    Utils.SHARED_PLUGIN_NAMESPACE,
+    SHADER_PLUGIN_DATA_KEY
+  );
+  if (shaderCode && shaderCode.length != 0) {
+    let nodeWithFills: MinimalFillsMixin = node as MinimalFillsMixin;
+
+    if (nodeWithFills) {
+      nodeWithFills.fills = [
```
[the remark this answers is not in the document]
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/shader.ts]: Per discussion offline, remove shader image preview when clear shader. And if it is text, we will set the text color to the solid pink if there is no background afterwards.

**Pull request merged**